### PR TITLE
 Fixes for GCC compatibility and Linux platform warnings

### PR DESCRIPTION
**Pull request by PatriceJiang**
Re: https://github.com/cocos/3d-tasks/issues/15654

### Changelog

- Fixed compilation errors related to GCC, mainly through compatibility changes for templates.
- Ignored certain compilation warnings on Linux platforms.
- Linux platform now defaults to using GCC, so developers usually don't need to install clang separately.

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/base/StringPool.h [hunk 1]
```diff
@@ -35,7 +35,7 @@
 
 namespace cc {
 
-namespace {
+namespace impl {
```
Comment (minggo) [inline native/cocos/base/StringPool.h]: Why need a namespace name? This name is easy to conflict.

File: native/cmake/predefine.cmake [hunk 2]
```diff
@@ -116,7 +116,13 @@ if("$ENV{COCOS_ENGINE_DEV}" EQUAL "1")
...
     
     if(APPLE)
-        set(WERROR_FLAGS "${WERROR_FLAGS} -Wno-deprecated-declarations")
+        set(WERROR_FLAGS " ${WERROR_FLAGS} -Wno-deprecated-declarations")
+    elseif(LINUX)
+        set(WERROR_FLAGS " ${WERROR_FLAGS} -Wno-nullability-completeness -Wno-deprecated-declarations")
+    endif()
+
+    if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+        set(WERROR_FLAGS " ${WERROR_FLAGS} -Wno-invalid-offsetof")
```
Comment (minggo) [inline native/cmake/predefine.cmake:125]: It is better to do it only in linux.

File: native/cocos/base/StringPool.h (hunk 1 above)
Comment (PatriceJiang) [reply native/cocos/base/StringPool.h]: gcc reports error here
```
/home/pt/Github/engine/native/cocos/base/StringPool.h: In instantiation of ‘class cc::StringPool<false>’:

/home/pt/Github/engine/native/cocos/renderer/frame-graph/PassInsertPointManager.h:53:16:   required from here

/home/pt/Github/engine/native/cocos/base/StringPool.h:55:7: error: ‘cc::StringPool<false>’ has a field ‘cc::StringPool<false>::_stringToHandles’ whose type uses the anonymous namespace [-Werror=subobject-linkage]
```

File: native/cmake/predefine.cmake (hunk 2 above)
Comment (PatriceJiang) [reply native/cmake/predefine.cmake:125]: Other platforms, such as CDC, may also use the GCC compiler.

File: native/cocos/base/StringPool.h (hunk 1 above)
Comment (PatriceJiang) [reply native/cocos/base/StringPool.h]: The warning `-subobject-linkage` has been ignored.

Comment (PatriceJiang) [reply native/cocos/base/StringPool.h]: The definitions of StringEqual and StringHash have been moved to the inside of the StringPool class.

**Review by minggo — APPROVED** [review]

**Pull request merged**